modules/host/update.go

```diff
@@ -0,0 +1,117 @@
+// determine the random seed. Create proof must be under a host and state lock.
+func (h *Host) createStorageProof(obligation contractObligation, heightForProof consensus.BlockHeight) (err error) {
+	fullpath := filepath.Join(h.hostDir, obligation.path)
+	file, err := os.Open(fullpath)
+	if err != nil {
+		return
+	}
+	defer file.Close()
+
+	segmentIndex, err := h.state.StorageProofSegment(obligation.id)
+	if err != nil {
+		return
+	}
+	numSegments := crypto.CalculateSegments(obligation.fileContract.FileSize)
+	base, hashSet, err := crypto.BuildReaderProof(file, numSegments, segmentIndex)
+	if err != nil {
+		return
+	}
+
+	sp := consensus.StorageProof{obligation.id, base, hashSet}
+
+	// Create and send the transaction.
+	id, err := h.wallet.RegisterTransaction(consensus.Transaction{})
+	if err != nil {
+		return
+	}
+	_, _, err = h.wallet.AddStorageProof(id, sp)
+	if err != nil {
+		return
+	}
+	t, err := h.wallet.SignTransaction(id, true)
+	if err != nil {
+		return
+	}
+	err = h.tpool.AcceptTransaction(t)
+	if err != nil {
+		if consensus.DEBUG {
+			panic(err)
+		}
+	}
+
+	return
+}
+
+// update grabs all of the blocks that have appeared since the last update and
+// submits any necessary storage proofs.
+func (h *Host) update() {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+
+	// Get the blocks since the recent update.
+	_, appliedBlockIDs, err := h.state.BlocksSince(h.latestBlock)
+	if err != nil {
+		// The host has somehow desynchronized.
+		if consensus.DEBUG {
+			panic(err)
+		}
+	}
+	if len(appliedBlockIDs) == 0 {
+		return
+	}
+	h.latestBlock = appliedBlockIDs[len(appliedBlockIDs)-1]
+
+	// Check the applied blocks and see if any of the contracts we have are
+	// ready for storage proofs.
+	for _, blockID := range appliedBlockIDs {
+		height, exists := h.state.HeightOfBlock(blockID)
+		if consensus.DEBUG {
+			if !exists {
+				panic("a block returned by BlocksSince doesn't appear to exist")
+			}
+		}
+
+		for _, obligation := range h.obligationsByHeight[height] {
+			// Submit a storage proof for the obligation.
+			err := h.createStorageProof(obligation, h.state.Height())
+			if err != nil {
+				fmt.Println(err)
+				return
+			}
+
+			// Delete the obligation.
+			fullpath := filepath.Join(h.hostDir, obligation.path)
+			stat, err := os.Stat(fullpath)
+			if err != nil {
+				fmt.Println(err)
+				return
+			}
+			h.deallocate(uint64(stat.Size()), obligation.path) // TODO: file might actually be the wrong size.
+
+			delete(h.obligationsByID, obligation.id)
+		}
+		delete(h.obligationsByHeight, height)
+	}
+}
+
+// threadedConsensusListen listens to a channel that's subscribed to the state
+// and updates every time the consensus set changes. When the consensus set
+// changes, the host checks if there are any storage proofs that need to be
+// submitted and submits them.
+func (h *Host) threadedConsensusListen(consensusChan <-chan struct{}) {
+	for _ = range consensusChan {
+		h.update()
+	}
+}
```

Comment:
you know, if every module had an update function...
and if the State was a module...
we could make an `Updater` interface...
and the State could takes a slice of `Updaters`...
and call `Update` on them itself instead of sending down a bunch of channels.